teps/0021-results-api.md

```diff
@@ -0,0 +1,772 @@
+name easily and efficiently. However, Kubernetes' support for listing/querying
+across many custom resources is severely lacking.
+
+Kubernetes allows users to request filtered lists of API resources using
+[field selectors and label selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/).
+For custom resources the only support field selectors are name and namespace.
+Label selectors are likewise limited, since we'd have to store any queryable
+information about a Run in a label for the Run to be able to query for it, and
+this grows large quickly. Together, these limitations make it pretty much
+impossible to query for anything actually useful today.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Define a Result API spec to store and query Tekton results.
+  - Tekton results are any data about or related to TaskRuns or PipelineRuns. At
+    minimum this includes the full TaskRun and PipelineRun status/spec, but may
+    include other data about the Runs (events that triggered it, artifacts
+    produced, post-run tasks, etc.)
+- Decouple long-term Tekton result storage away from runtime storage in etcd.
+- Give clients flexibility to embed additional metadata associated with Tekton
+  Task/PipelineRuns.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Log exporting - we will rely on
+  [existing projects](https://landscape.cncf.io/category=logging&format=card-mode&grouping=category)
+  to handle exporting of logs to remote systems.
+- Defining a system to automatically clean up completed Task/PipelineRuns.
+- Define specific authentatication/authorization mechanisms (this will likely be
+  implementation specific).
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+- Results integrations should be an optional add-on on top of Tekton. Tekton
+  Pipelines should be able to operate normally without the presense of a Result
+  server.
+- Result API Server hosts should be able to store multi-tenant results and share
+  the server among multiple namespaces or clusters.
+- Service providers can choose to provide their own controller and server
+  implementations to upload Results with additional service specific metadata or
+  implementation specific validation / field formats, so long as it conforms to
+  the Results API.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Result API Spec
+
+Tekton should document a versioned gRPC API (with
+[JSON/HTTP bindings](https://cloud.google.com/endpoints/docs/grpc/transcoding))
```

Comment:
maybe link to the REST alternative below (I went looking for it right after i read this section)